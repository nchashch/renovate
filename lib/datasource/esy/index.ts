import { logger } from '../../logger';
import got from '../../util/got';
import { PkgReleaseConfig, ReleaseResult, Release } from '../common';

export async function getPkgReleases({
  lookupName,
}: PkgReleaseConfig): Promise<ReleaseResult> {
  const path: string = `/repos/ocaml/opam-repository/contents/packages/${lookupName}/`;
  const baseUrl: string = 'https://api.github.com';
  const packageUrl: string = baseUrl + path;
  console.log(packageUrl);
  let res: any = await got(packageUrl, {
    hostType: 'esy',
    json: true,
  });
  // TODO: Read homepage, name, and maybe other metadata from /packages/${lookupName}/${lookupName.1.2.3}/opam
  // where 1.2.3 can be any existing version, check if there is a mechanism for deprication
  const versions: string[] = res.body.map(file =>
    file.name.substring(file.name.indexOf('.') + 1)
  );
  const releases: Release[] = versions.map(version => ({ version }));
  const result: ReleaseResult = { releases };
  return result;
}
