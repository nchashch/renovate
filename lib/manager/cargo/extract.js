const versioning = require('../../versioning');

const semver = versioning('semver');

module.exports = {
  extractPackageFile,
};

/* Quote from TOML spec. */
/*
Inline tables are intended to appear on a single line. No newlines are
allowed between the curly braces unless they are valid within a value.
Even so, it is strongly discouraged to break an inline table onto
multiples lines. If you find yourself gripped with this desire, it
means you should be using standard tables.
*/

// TODO: Treat invalid versions gracefully
// TODO: Refactor
// TODO: Find and fix all corner cases
function extractPackageFile(content, fileName) {
  logger.trace(`cargo.extractPackageFile(${fileName})`);
  const deps = [];
  let lineNumber = 0;
  const lines = content.split('\n');
  for (const line of lines) {
    const match = line.match('dependencies[.]');
    if (match) {
      const dep = {};
      let input = match.input;
      input = input.replace(/[[\]]+/g, '');
      input = input.split('.');
      dep.depName = input[input.length - 1];
      // Record the line number of [*dependencies.<depName>] expression
      dep.lineNumber = lineNumber;
      let name = '';
      let i = 1;
      while (name !== 'version') {
        const field = lines[lineNumber + i].split('=');
        name = field[0].trim();
        if (name === 'version') {
          dep.currentValue = field[1]
            .trim()
            .replace(/['"]+/g, '')
            .trim();
        }
        i += 1;
      }
      deps.push(dep);
    } else if (line.match('dependencies')) {
      let i = 1;
      while (
        lineNumber + i < lines.length &&
        lines[lineNumber + i].trim()[0] !== '['
      ) {
        let dep = null;
        const l = lines[lineNumber + i];
        if (l.match('{*}')) {
          dep = getInlineDep(l);
        } else {
          dep = getDep(l);
        }
        if (dep) {
          dep.lineNumber = lineNumber + i;
          deps.push(dep);
        }
        i += 1;
      }
    }
    lineNumber += 1;
  }
  if (!deps.length) {
    return null;
  }
  for (let i = 0; i < deps.length; i += 1) {
    if (!semver.isValid(deps[i].currentValue)) {
      deps[i].skipReason = 'unknown-version';
    }
  }
  return { deps };
}

function getDep(line) {
  const entry = line.split('=');
  const dep = {};
  const depName = entry[0];
  if (depName) {
    const currentValue = entry[1]
      .trim()
      .replace(/['"]+/g, '')
      .trim();
    dep.currentValue = currentValue;
    dep.depName = depName;
    return dep;
  }
  return null;
}

// NOTE: Some corner cases, like nested {} might not be covered
// TODO: Find and fix corner cases
function getInlineDep(line) {
  const dep = {};
  let content = line.split('=');
  dep.depName = content[0];
  content = content.slice(1);
  content = content.join('=');
  content = content.trim().replace(/[{}]+/g, '');
  const fields = content.split(',');
  for (let i = 0; i < fields.length; i += 1) {
    const field = fields[i].split('=');
    const name = field[0].trim();
    const value = field[1].trim();
    if (name === 'version') {
      dep.currentValue = value.replace(/['"]+/g, '').trim();
      return dep;
    }
  }
  return null;
}
