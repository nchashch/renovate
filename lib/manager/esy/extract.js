module.exports = {
  extractPackageFile,
};

// NOTE: It looks like esy uses the same dependencies, devDependencies,
// buildDependencies fields in package.json as npm. But package.json also can
// have esy field. Also it uses both npm and opam, so a new datasource might
// be necessary.
function extractPackageFile(content) {
  logger.trace(`extractPackageFile()`);
  try {
    const doc = JSON.parse(content);
    if (doc.esy) {
      logger.info('esy field is present');
    }
  } catch (err) {
    return {};
  }
  return {};
}
