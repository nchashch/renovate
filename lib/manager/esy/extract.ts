import { logger } from '../../logger';
import { PackageDependency, PackageFile } from '../common';
import { EsySection } from './types';

export function extractPackageFile(
  content: string,
  fileName: string
): PackageFile | null {
  logger.trace(`extractPackageFile()`);
  try {
    const doc = JSON.parse(content);
    if (doc.esy) {
      logger.info('esy field is present');
    }
    const deps = [
      ...extractFromSection(doc, 'dependencies'),
      ...extractFromSection(doc, 'devDependencies'),
      ...extractFromSection(doc, 'buildDependencies'),
    ];
    if (deps.length === 0) {
      return null;
    }
    return { deps };
  } catch (err) {
    return null;
  }
}

function extractFromSection(
  parsedContent: EsySection,
  section: keyof EsySection
): PackageDependency[] {
  const deps: PackageDependency[] = [];
  const sectionContent = parsedContent[section];
  if (!sectionContent) {
    return [];
  }
  Object.keys(sectionContent).forEach(depName => {
    const parsedDepName = parseDepName(depName);
    const currentValue = sectionContent[depName];
    if (currentValue) {
      const npmScope = parsedDepName.npmScope;
      const dep: PackageDependency = {
        depName,
        currentValue,
        depType: section,
        datasource: npmScope === 'opam' ? 'opam' : 'npm',
      };
      deps.push(dep);
    }
  });
  return deps;
}

export function parseDepName(depName) {
  const strs = depName.split('/');
  if (strs.length === 1) {
    return {
      depName: strs[0],
    };
  }
  if (strs.length === 2) {
    if (strs[0][0] === '@') {
      strs[0] = strs[0].substring(1);
    } else {
      return null;
    }
    return {
      npmScope: strs[0],
      depName: strs[1],
    };
  }
  return null;
}
