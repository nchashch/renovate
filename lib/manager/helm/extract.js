const upath = require('upath');
const yaml = require('js-yaml');
const is = require('@sindresorhus/is');

module.exports = {
  extractPackageFile,
};

async function extractPackageFile(content, fileName) {
  let chartName;
  try {
    const baseDir = upath.parse(fileName).dir;
    const chartFileName = upath.join(baseDir, 'Chart.yaml');
    const chart = await platform.getFile(chartFileName);
    if (!chart) {
      logger.warn('Failed to read Chart.yaml');
    }
    const doc = yaml.safeLoad(chart);
    chartName = doc.name;
  } catch (err) {
    logger.warn('Failed to parse Chart.yaml');
  }
  logger.info(`Chart name: ${chartName}`);
  logger.trace('helm.extractPackageFile()');
  let deps = [];
  try {
    const doc = yaml.safeLoad(content);
    if (doc && is.array(doc.dependencies)) {
      deps = doc.dependencies.map(dep => {
        const res = {
          depName: dep.name,
          currentValue: dep.version,
          helmRepository: dep.repository,
        };
        const url = new URL(dep.repository);
        if (url.protocol === 'file:') {
          res.skipReason = 'local-dependency';
        }
        return res;
      });
    }
  } catch (err) {
    logger.warn(
      { err, fileName },
      'extractPackageFile failed to parse requirements.yaml file'
    );
  }
  if (deps.length === 0) {
    logger.debug(
      { fileName },
      "extractPackageFile didn't extract any dependencies"
    );
    return null;
  }
  const res = {
    deps,
    datasource: 'helm',
  };
  if (chartName) {
    res.chartName = chartName;
  }
  return res;
}
