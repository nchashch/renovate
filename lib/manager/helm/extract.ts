import is from '@sindresorhus/is';
import upath from 'upath';
import yaml from 'js-yaml';

import { logger } from '../../logger';
import { PackageFile, PackageDependency } from '../common';

export async function extractPackageFile(
  content: string,
  fileName: string
): Promise<PackageFile> {
  try {
    const baseDir = upath.parse(fileName).dir;
    const chartFileName = upath.join(baseDir, 'Chart.yaml');
    const chart = await platform.getFile(chartFileName);
    if (!chart) {
      logger.warn('Failed to read Chart.yaml');
    }
  } catch (err) {
    logger.warn('Failed to parse Chart.yaml');
  }
  logger.trace('helm.extractPackageFile()');
  let deps = [];
  try {
    const doc = yaml.safeLoad(content);
    if (doc && is.array(doc.dependencies)) {
      deps = doc.dependencies.map(dep => {
        const res: PackageDependency = {
          depName: dep.name,
          currentValue: dep.version,
          registryUrls: [dep.repository],
        };
        const url = new URL(dep.repository);
        if (url.protocol === 'file:') {
          res.skipReason = 'local-dependency';
        }
        return res;
      });
    }
  } catch (err) {
    logger.warn(
      { err, fileName },
      'extractPackageFile failed to parse requirements.yaml file'
    );
  }
  if (deps.length === 0) {
    logger.debug(
      { fileName },
      "extractPackageFile didn't extract any dependencies"
    );
    return null;
  }
  const res = {
    deps,
    datasource: 'helm',
  };
  return res;
}
