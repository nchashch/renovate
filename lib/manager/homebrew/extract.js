const semver = require('../../versioning/semver');

module.exports = {
  extractPackageFile,
};

function extractPackageFile(content) {
  /*
    1. match "class depName < Formula"
    2. extract depName
    3. extract url field
    4. extract sha256 field
  */
  const depName = extractDepName(content);
  if (!depName) {
    logger.debug('Invalid class definition');
    return null;
  }
  const url = extractUrl(content);
  if (!url) {
    logger.debug('Invalid URL field');
  }
  const currentValue = parseCurrentValue(url);
  let skipReason;
  if (!currentValue) {
    logger.debug('Error: Unsupported URL field');
    skipReason = 'unsupported-url';
  }
  // const sh256 = extractSha256(content);
  const dep = {
    depName,
    url,
    currentValue,
  };
  if (skipReason) {
    dep.skipReason = skipReason;
  }
  const deps = [dep];
  return { deps };
}

function extractUrl(content) {
  const urlRegExp = /(^|\s)url(\s)/;
  let i = content.search(urlRegExp);
  if (isSpace(content[i])) {
    i += 1;
  }
  return parseUrl(i, content);
}

function parseCurrentValue(urlStr) {
  if (!urlStr) {
    return null;
  }
  try {
    const url = new URL(urlStr);
    if (url.hostname !== 'github.com') {
      return null;
    }
    const s = url.pathname.split('/');
    let currentValue;
    if (s[3] === 'archive') {
      currentValue = s[4];
      const targz = currentValue.slice(
        currentValue.length - 7,
        currentValue.length
      );
      if (targz === '.tar.gz') {
        currentValue = currentValue.substring(0, currentValue.length - 7);
      }
    } else if (s[3] === 'releases' && s[4] === 'download') {
      currentValue = s[5];
    }
    if (!currentValue) {
      return null;
    }
    if (!semver.isValid(currentValue)) {
      return null;
    }
    return currentValue;
  } catch (_) {
    return null;
  }
  return null;
}

function parseUrl(idx, content) {
  let i = idx;
  i += 'url'.length;
  i = skip(i, content, c => {
    return isSpace(c);
  });
  const c = content[i];
  if (c !== '"' && c !== "'") {
    return null;
  }
  i += 1;
  let j = i;
  j = skip(i, content, c => {
    return c !== '"' && c !== "'" && !isSpace(c);
  });
  const url = content.slice(i, j);
  return url;
}

function extractDepName(content) {
  const classRegExp = /(^|\s)class\s/;
  let i = content.search(classRegExp);
  if (isSpace(content[i])) {
    i += 1;
  }
  return parseClassHeader(i, content);
}

/* This function parses the "class depName < Formula" header
   and returns the depName and index of the character just after the header */
function parseClassHeader(i, content) {
  i = i + 'class'.length;
  i = skip(i, content, c => {
    return isSpace(c);
  });
  // Skip all non space and non '<' characters
  let j = skip(i, content, c => {
    return !isSpace(c) && c !== '<';
  });
  const depName = content.slice(i, j);
  i = j;
  // Skip spaces
  i = skip(i, content, c => {
    return isSpace(c);
  });
  if (content[i] === '<') {
    i += 1;
  } else {
    return null;
  } // Skip spaces
  i = skip(i, content, c => {
    return isSpace(c);
  });
  // Skip non-spaces
  j = skip(i, content, c => {
    return !isSpace(c);
  });
  if (content.slice(i, j) !== 'Formula') {
    return null;
  }
  return depName;
}

function skip(idx, content, cond) {
  let i = idx;
  while (i < content.length) {
    if (!cond(content[i])) {
      return i;
    }
    i += 1;
  }
  return i;
}

function isSpace(c) {
  return /\s/.test(c);
}
