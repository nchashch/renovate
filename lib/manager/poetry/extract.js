const toml = require('toml');
const semver = require('../../versioning/poetry');

module.exports = {
  extractPackageFile,
};

function extractPackageFile(content, fileName) {
  logger.trace(`poetry.extractPackageFile(${fileName})`);
  let pyprojectfile;
  try {
    pyprojectfile = toml.parse(content);
  } catch (err) {
    logger.debug({ err }, 'Error parsing pyproject.toml file');
    return null;
  }
  const dependencies = pyprojectfile.tool.poetry.dependencies;
  const devDeps = pyprojectfile.tool.poetry['dev-dependencies'];
  const extras = pyprojectfile.tool.poetry.extras;
  const allDeps = { ...dependencies, ...devDeps, ...extras };
  const deps = [];
  let skipReason;
  Object.keys(allDeps).forEach(depName => {
    let currentValue = allDeps[depName];
    if (typeof currentValue !== 'string') {
      const version = allDeps[depName].version;
      const path = allDeps[depName].path;
      if (version) {
        currentValue = version;
      } else if (path) {
        currentValue = '';
        skipReason = 'path-dependency';
      } else {
        currentValue = '';
        skipReason = 'multiple-constraint-dep';
      }
    }
    const dep = {
      depName,
      currentValue,
      datasource: 'pypi',
    };
    if (skipReason) {
      dep.skipReason = skipReason;
    } else if (!semver.isValid(dep.currentValue)) {
      dep.skipReason = 'unknown-version';
    }
    deps.push(dep);
  });
  return { deps };
}
