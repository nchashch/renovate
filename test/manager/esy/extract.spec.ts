import { readFileSync } from 'fs';
import { extractPackageFile } from '../../../lib/manager/esy/extract';

const package1json = readFileSync(
  'test/manager/esy/_fixtures/package.1.json',
  'utf8'
);
const package2json = readFileSync(
  'test/manager/esy/_fixtures/package.2.json',
  'utf8'
);

describe('lib/manager/esy/extract', () => {
  describe('extractPackageFile()', () => {
    it('returns null for empty', async () => {
      const content = '';
      const fileName = '';
      expect(await extractPackageFile(content, fileName)).toBeNull();
    });
    it('parses a simple JSON object', async () => {
      const content = package1json;
      const fileName = 'package.json';
      expect(await extractPackageFile(content, fileName)).not.toBeNull();
      expect(await extractPackageFile(content, fileName)).toMatchSnapshot();
    });
    it('parses a simple JSON object', async () => {
      const content = package2json;
      const fileName = 'package.json';
      expect(await extractPackageFile(content, fileName)).not.toBeNull();
      expect(await extractPackageFile(content, fileName)).toMatchSnapshot();
    });
    it('returns null if there are no dependencies', async () => {
      const content = `{
  "name": "hello-reason",
  "version": "0.1.0",
  "description": "Example Reason Esy Project",
  "license": "MIT" }`;
      const fileName = 'package.json';
      const res = await extractPackageFile(content, fileName);
      expect(res).toBeNull();
    });
    it('fails if there are invalid depNames', async () => {
      const content = `{
  "name": "hello-reason",
  "version": "0.1.0",
  "description": "Example Reason Esy Project",
  "license": "MIT",
  "dependencies": {
    "@opam/foo/bar": "1.2.3"
  }
}`;
      const fileName = 'package.json';
      const res = await extractPackageFile(content, fileName);
      expect(res).toBeNull();
    });
    it('fails if there are invalid depNames', async () => {
      const content = `{
  "name": "hello-reason",
  "version": "0.1.0",
  "description": "Example Reason Esy Project",
  "license": "MIT",
  "dependencies": {
    "opam/foo": "1.2.3"
  }
}`;
      const fileName = 'package.json';
      const res = await extractPackageFile(content, fileName);
      expect(res).toBeNull();
    });
  });
});
