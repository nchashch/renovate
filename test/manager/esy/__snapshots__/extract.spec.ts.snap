// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`lib/manager/esy/extract extractPackageFile() parses a simple JSON object 1`] = `
Object {
  "deps": Array [
    Object {
      "currentValue": "*",
      "datasource": "opam",
      "depName": "@opam/dune",
      "depType": "dependencies",
    },
    Object {
      "currentValue": "*",
      "datasource": "opam",
      "depName": "@opam/lambda-term",
      "depType": "dependencies",
    },
    Object {
      "currentValue": "^4.0.0",
      "datasource": "opam",
      "depName": "@opam/lwt",
      "depType": "dependencies",
    },
    Object {
      "currentValue": "~4.6.0",
      "datasource": "npm",
      "depName": "ocaml",
      "depType": "dependencies",
    },
    Object {
      "currentValue": "^3.0.3",
      "datasource": "opam",
      "depName": "@opam/merlin",
      "depType": "devDependencies",
    },
    Object {
      "currentValue": "~4.6.0",
      "datasource": "npm",
      "depName": "ocaml",
      "depType": "devDependencies",
    },
  ],
}
`;

exports[`lib/manager/esy/extract extractPackageFile() parses a simple JSON object 2`] = `
Object {
  "deps": Array [
    Object {
      "currentValue": "*",
      "datasource": "opam",
      "depName": "@opam/dune",
      "depType": "dependencies",
    },
    Object {
      "currentValue": "*",
      "datasource": "npm",
      "depName": "@reason-native/console",
      "depType": "dependencies",
    },
    Object {
      "currentValue": "*",
      "datasource": "npm",
      "depName": "@reason-native/pastel",
      "depType": "dependencies",
    },
    Object {
      "currentValue": "*",
      "datasource": "npm",
      "depName": "@reason-native/rely",
      "depType": "dependencies",
    },
    Object {
      "currentValue": ">= 3.4.0 < 3.6.0",
      "datasource": "npm",
      "depName": "@esy-ocaml/reason",
      "depType": "dependencies",
    },
    Object {
      "currentValue": "*",
      "datasource": "npm",
      "depName": "refmterr",
      "depType": "dependencies",
    },
    Object {
      "currentValue": "~4.6.0",
      "datasource": "npm",
      "depName": "ocaml",
      "depType": "dependencies",
    },
    Object {
      "currentValue": "*",
      "datasource": "opam",
      "depName": "@opam/merlin",
      "depType": "devDependencies",
    },
    Object {
      "currentValue": "~4.6.0",
      "datasource": "npm",
      "depName": "ocaml",
      "depType": "devDependencies",
    },
    Object {
      "currentValue": "*",
      "datasource": "opam",
      "depName": "@opam/odoc",
      "depType": "devDependencies",
    },
  ],
}
`;
